File: Backend/src/main/java/com/imageprocessing/service/ImageService.java
package com.imageprocessing.service;

import com.imageprocessing.dto.*;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;

public interface ImageService {

    UploadResponse uploadImage(
            MultipartFile file)
            throws IOException;


    BrightnessResponse adjustBrightnessresponse(
            String imageId,
            int brightness
    );

    ContrastResponse adjustContrastresponse(
            String imageId,
            double contrast);

    BlurResponse blurImageresponse(
            String imageId,
            int kernelSize);

    SharpnessResponse enhanceSharpness(
            String imageId,
            double factor);

    RotateResponse rotateImage(
            String imageId,
            int angle);

    FlipResponse flipImage(
            String imageId,
            String direction);

    BackgroundRemovalResponse removeBackground(
            String imageId,
            int threshold);

    GrayscaleResponse convertToGrayscale(
            String imageId);

    ShapeDetectionResponse detectShape(
            String imageId);

    ZoomResponse zoomImage(
            String imageId,
            double factor);

    LayerResponse layerImages(
            String backgroundImageId,
            MultipartFile foregroundFile,
            int xOffset,
            int yOffset)
            throws IOException;

    UndoResponse undo(
            String imageId
    );

    byte[] downloadImage(
            String imageId)
            throws IOException;


}

File: frontend/src/services/imageService.js
import axios from "axios";

const API =
    "http://localhost:8080/api";

export const uploadImage =
    async (file) => {

        const formData =
            new FormData();

        formData.append(
            "file",
            file
        );

        const response =
            await axios.post(
                `${API}/upload`,
                formData,
                {
                    headers: {
                        "Content-Type":
                            "multipart/form-data"
                    }
                }
            );

        return response.data;
    };


export const adjustBrightness =
    async (imageId, brightness) =>
        (await axios.post(
            `${API}/${imageId}/brightness?brightness=${brightness}`
        )).data;

export const adjustContrast =
    async (imageId, contrast) =>
        (await axios.post(
            `${API}/${imageId}/contrast?contrast=${contrast}`
        )).data;

export const blurImage =
    async (imageId, kernelsize) =>
        (await axios.post(
            `${API}/${imageId}/blur?kernelsize=${kernelsize}`
        )).data;

export const enhanceSharpness =
    async (imageId, factor) =>
        (await axios.post(
            `${API}/${imageId}/sharpness?factor=${factor}`
        )).data;

export const rotateImage =
    async (imageId, angle) =>
        (await axios.post(
            `${API}/${imageId}/rotate?angle=${angle}`
        )).data;

export const flipImage =
    async (imageId, direction) =>
        (await axios.post(
            `${API}/${imageId}/flip?direction=${direction}`
        )).data;

export const grayscaleImage =
    async (imageId) =>
        (await axios.post(
            `${API}/${imageId}/grayscale`
        )).data;

export const removeBackground =
    async (imageId, threshold) =>
        (await axios.post(
            `${API}/${imageId}/background-remove?threshold=${threshold}`
        )).data;

export const zoomImage =
    async (imageId, factor) =>
        (await axios.post(
            `${API}/${imageId}/zoom?factor=${factor}`
        )).data;

export const undoImage = async (imageId) => {
    const response =
        await axios.post(
            `${API}/${imageId}/undo`
        );

    return response.data;
};

export const detectShape =
    async (imageId) =>
        (await axios.post(
            `${API}/${imageId}/shape-detection`
        )).data;

export async function downloadImage(imageId) {

    const response = await fetch(
        `http://localhost:8080/api/download/${imageId}`
    );

    if (!response.ok) {
        throw new Error("Failed to download image");
    }

    const blob = await response.blob();

    const url = window.URL.createObjectURL(blob);

    const link = document.createElement("a");

    link.href = url;
    link.download = `image-${imageId}.png`;

    document.body.appendChild(link);

    link.click();

    link.remove();

    window.URL.revokeObjectURL(url);
}

export const layerImages = async (
    backgroundImageId,
    foregroundFile,
    xOffset,
    yOffset
) => {

    const formData =
        new FormData();

    formData.append(
        "backgroundImageId",
        backgroundImageId
    );

    formData.append(
        "foregroundFile",
        foregroundFile
    );

    formData.append(
        "xOffset",
        xOffset
    );

    formData.append(
        "yOffset",
        yOffset
    );

    const response =
        await axios.post(
            `${API}/layer`,
            formData,
            {
                headers: {
                    "Content-Type":
                        "multipart/form-data"
                }
            }
        );

    return response.data;
};